glsl.meta.slang_misc_glsl #3692, review

ArielG-NV opened the pull request.

ArielG-NV committed:
did not implement hlsl for many functions since they would require  raytracing level workarounds due to a more object oriented style rather than location numbers.

did not implement subpass yet since this will require overhaul of texture internally in quite a few areas

ArielG-NV committed:
fully add miscellaneous functions and tests:
Many functions invocation-group & invocation-control functions are partially removed from vulkan or deprecated (for a number of years). These functions use alternative methods of implementation to ensure they function at a minimum level of standard of 'working as expected'

noise functions are fully deprecated.

subpass images are missing an implementation in Slang, these were added

shader-memory-control functions are mostly working from GLSL->Vulkan, memoryBarrierAtomicCounter() is 'removed', although due to including atomic_counter (GL_EXT_vulkan_glsl_relaxed compiler extension), the function is kept and wrapped as a memoryBarrierBuffer() to exhibit equivalent functionality

Fragment Processing should be fully working.

Geometry shader functionality as a whole is unimplemented and will need to be added before attempting to test geometry shader functions

ArielG-NV committed:
syntax adjustments

fix syntax and remove unused geometry shader functions (untested as well)

Some syntax adjustments.

ArielG-NV committed:
misc syntax adjustment and tryijng to get interpolate to work

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
Merge branch 'glsl.meta.slang_miscGLSL_#3692' of https://github.com/ArielG-NV/slang into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
fix regression of fallthrough switch case

ArielG-NV committed:
add code changes which enable a working error system and variable passing system for when functions require 'input' decorated variables

add code changes which enable a working error system and variable passing system for when functions require 'input' decorated variables

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
general fixes for implementation before review

fix compile warning & add missing modifiers & adjust test

int vs uint comparison

add side effect modifiers

removing junk added to test to ensure testing is working

likley will break falcor & will have to change slang.h interactions, but I will do that when I have a change to test with falcor

ArielG-NV committed:
prepare code to be ready for review

grammer & style

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_miscGLSL_#3692

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+        return spirv_asm
+        {
+            OpCapability InterpolationFunction;
+            OpExtInst $$float result glsl450 78 &interpolant $offset
```
I am curious what the magic number 78 means here.

ArielG-NV replied:
for some reason interpolation functions don't always compile correctly on the glslang side of things, 78 is the number of `interpolateAtOffset` for the glsl450 extended capability set of `OpExtInst`. 

I will try to see if `InterpolateAtOffset` works again, although I don't believe it does work

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+    case hlsl:
+    {
+        GroupMemoryBarrier();
+        DeviceMemoryBarrier();
```
I don't think "DeviceMemoryBarrier()" is needed for "barrier".
The spec says that it is for the shard memory and there is a different APIs for the memory barrier.
```
A barrier() affects control flow but only synchronizes memory accesses to shared variables and
tessellation control output variables. For other memory accesses, it does not ensure that values
written by one invocation prior to a given static instance of barrier() can be safely read by other
invocations after their call to the same static instance of barrier(). To achieve this requires the use
of both barrier() and a memory barrier.
```

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Invocation AcquireRelease|UniformMemory
```
Shouldn't this have
```
AtomicCounterMemory
```

ArielG-NV replied:
AtomicCounterMemory is invalid in vulkan. Slang converts all atomic_uint's as per GL_EXT_vulkan_glsl_relaxed into a storage buffer. I will add this ext to the comment to give better scope of what we are doing.

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+        spirv_asm
+        {
+            OpCapability Shader;
+            OpMemoryBarrier Invocation AcquireRelease|WorkgroupMemory|ImageMemory|UniformMemory
```
Should this be limited to Workgroup?

ArielG-NV replied:
as per spec, the following block within a shader invocation scope:
* memoryBarrier
* memoryBarrierAtomicCounter
* memoryBarrierBuffer
* memoryBarrierImage

the following block within workgroup memory scope:
* memoryBarrierShared
* groupMemoryBarrier

jkwak-work replied:
That is right.
So you shouldn't have WorkgroupMemory here.
The keyword in the word, "WorkgroupMemory", is "Memory" not "Workgroup".
groupshared memory is available only for compute shader and it means WorkgroupMemory should be used only for "memoryBarrierShared" and "groupMemoryBarrier". Both functions are only for compute shaders.

ArielG-NV replied:
shaderplayground compiles `memoryBarrier` to `OpMemoryBarrier %uint_1 %uint_3400`

this is the same as: 
Device scope, memory { Acquire Release | uniform | workgroup | atomic_counter|imageMemory}

ArielG-NV replied:
Device scope is clearly going against OpenGL Specification of `invocation`, but the memory semantics targeted does include all relevant memory types, and is likely correct

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +6803,720 @@ __glsl_version(430) [require(glsl)]
+    {
+        spirv_asm
+        {
+            OpMemoryBarrier Workgroup AcquireRelease|WorkgroupMemory|ImageMemory|UniformMemory
```
I don't think this should be limited to Workgroup.
The document describes "WorkgroupMemory" as,
```
Apply the memory-ordering constraints to Workgroup Storage Class memory.
```
This is not about the threads in workgroup.
It is about the memory types that are for workgroups such as shared memory.

ArielG-NV replied:
let me see what shader-playground compiles the function as

ArielG-NV replied:
shaderplayground compiles `OpMemoryBarrier %uint_2 %uint_3400`
2 = WorkGroup Memory scope

3400 = 0x8|0x40|0x100|0x400|0x800
Acquire Release | uniform | workgroup | atomic_counter|imageMemory

hmm, looks like AtomicCounterMemory is used

ArielG-NV replied:
if we try to use atomicMemory with atomicStorage capability (Required):
```
Validation: error 0: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08737 ] | MessageID = 0xa5625282 | vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
Capability AtomicStorage is not allowed by Vulkan 1.2 specification (or requires extension)
  OpCapability AtomicStorage
. The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must adhere to the validation rules described by the Validation Rules within a Module section of the SPIR-V Environment appendix (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08737)

Validation: error 0: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08739 ] | MessageID = 0x605314fa | vkCreateComputePipelines(): pCreateInfos[0].stage SPIR-V has Capability (Unhandled OpCapability) declared, but this is not supported by Vulkan. The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must not declare any capability that is not supported by the API, as described by the Capabilities section of the SPIR-V Environment appendix (https://vulkan.lunarg.com/doc/view/1.3.275.0/windows/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08739)
```

The reason glslang compiles and we don't is because glslang does not emit `AtomicStorage` capability, which is required by SPIR-V spec. 
glslang should fail the compile, but it does not, the flag for `AtomicStorage` is likely ignored since the capability is not toggled

jkwak-work replied:
I don't know why you are talking about AtomicCounterMemory here.
But I take back what I said.
I think "groupMemoryBarrier()" should keep WorkgrouMemory.

ArielG-NV replied:
I brought up `AtomicCounterMemory` since I decided to look at what glslang produces for glsl->SPIR-V to see what is the correct answer

and it turns out glslang produces incorrect SPIR-V which really should not work

jkwak-work commented on source/slang/slang-ast-modifier.h:
```diff
@@ -243,6 +243,11 @@ class GLSLLocationLayoutModifier : public GLSLParsedLayoutModifier
     SLANG_AST_CLASS(GLSLLocationLayoutModifier)
 };
 
+class GLSLInputAttachmentIndexAttribute : public GLSLParsedLayoutModifier
```
The name of the class should follow the pattern on the other names.
```
GLSLXXXLayoutModifier
```

jkwak-work commented on source/slang/slang-emit-glsl.cpp:
```diff
@@ -2645,33 +2652,45 @@ void GLSLSourceEmitter::emitVarDecorationsImpl(IRInst* varDecl)
             case kIROp_VulkanHitObjectAttributesDecoration:
+                postfix = toSlice("EXT");
                 prefix = toSlice("hitObjectAttribute");
                 postfix = toSlice("NV");
```
postfix is set twice here.

jkwak-work commented on source/slang/slang-emit-glsl.cpp:
```diff
@@ -2645,33 +2652,45 @@ void GLSLSourceEmitter::emitVarDecorationsImpl(IRInst* varDecl)
     for (auto decoration : varDecl->getDecorations())
```
This for-loop appears to be for ray-tracing judging by the comments at the beginning of the function.
You may need to separate your code for the image-attachment-index from this loop.

ArielG-NV replied:
the function was mostly used for raytracing, but it does have another modifier:  `IRGloballyCoherentDecoration`.

it may need to be seperated, but for now it is the only place which makes sense since it is an optional modifier.

About the comment, I don't know why it is there?

jkwak-work replied:
What I suggested is to not change the existing for-loop.
And add one more loop that does a similar thing.
```
for (auto decoration : varDecl->getDecorations())
{
... keep ray tracing stuff as it was ...
}
for (auto decoration : varDecl->getDecorations())
{
... add your new stuff here ...
}
```

ArielG-NV replied:
that sounds good, will do

jkwak-work commented on source/slang/slang-emit-hlsl.cpp:
```diff
@@ -276,6 +276,8 @@ void HLSLSourceEmitter::_emitHLSLTextureType(IRTextureTypeBase* texType)
     {
         case SLANG_TEXTURE_1D:		m_writer->emit("Texture1D");		break;
+        case SLANG_TEXTURE_SUBPASS: 
+            /* fallthrough */
```
Probably better to use a proper C++ syntax,
```
[[fallthrough]];
```

ArielG-NV replied:
we should probably also remove `/* fallthrough */` from the [coding style guide](https://github.com/shader-slang/slang/blob/master/docs/design/coding-conventions.md)

jkwak-work commented on source/slang/slang-ir-spirv-legalize.cpp:
```diff
@@ -360,7 +365,8 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         {
-            // If the texture has no format decoration, try to infer it from the type.
+            // If the texture has no format decoration, try to infer it from the type if not a subpass.
+            if (textureType->GetBaseShape() == SLANG_TEXTURE_SUBPASS) return;
```
Change comment to
```
If the texture has no format decoration and it is not a subpass, try to infer it from the type.
```
But it may look better a following,
```
        // If the texture has no format decoration, try to infer it from the type.
        if (format == ImageFormat::unknown)
        {
            // We cannot infer the format when the shape is subpass, because XXXX.
            if (textureType->GetBaseShape() == SLANG_TEXTURE_SUBPASS) return;
````

jkwak-work commented on source/slang/slang-ir-validate.cpp:
```diff
@@ -223,6 +223,13 @@ namespace Slang
+            // not be resolved yet, do not exception on these types.
+            if (inst->getOperand(0)->findDecoration<IRLayoutDecoration>())
+                return;
+            break;
```
I don't understand what this comment means.
Please make it easier to read and clear for the meaning.

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -333,6 +333,8 @@ struct ParameterBindingInfo
     size_t              space = 0;
     size_t              index = 0;
+    bool                hasInputAttachmentIndex = false;
+    size_t              inputAttachmentIndex = 0;
```
This looks a lot of waste in memory.
You may need to find another way to avoid adding them in ParameterBindingInfo.

It looks like ParameterBindingAndKindInfo inherits from ParameterBindingInfo with "kind" information.
You may need to explorer how it works and see if you can piggy-back on that pattern.

LayoutSize is trying to avoid adding a boolean by handling one special case, "infinite".
If you cannot avoid adding the member variables, I think that's what you can do at least.
You could make a new class like that to indicate a special case of not having the value.

ArielG-NV replied:
I will see if I can treat this type the same way we manage 'layout' uniforms. The cost of memory is 12bytes\*uniforms_in_shader\*2 (since an additional structure is used which also has additional data).

I will address this if I can in the next commit.

jkwak-work replied:
It is 16 byes not 12. There is a concept of "padding". In short, the address of a member variable must be aligned to its size. In your case bool will use 4 byes and there will be 4 byes padding before the next size_t.

ArielG-NV replied:
I think the best case is to use a magic number approach. The life times of these objects are either limited, or most will be culled out, meaning we only use 8bytes*Uniform for a portion of the compile.

the alternative solutions don't improve the problem:
1. we could make a struct which inherits. This would work, but then when adding the actual code it will be a bunch of if statements everywhere for handling 1 of the 23 cases we currently have.
2. we could do a Vulkan style extension system. This would work, but then we would have to add a RefPtr<> since Slang does not use dangling pointers which is already substantially larger than a 8 byte pointer.

if many exceptions are needed for this semantic system the extension system is likely preferred

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -351,6 +353,7 @@ struct UsedRangeSet : RefObject
     // Information on what ranges of "registers" have already
     // been claimed, for each resource type
     UsedRanges usedResourceRanges[kLayoutResourceKindCount];
+    UsedRanges usedResourceRangesInputAttachmentIndex[kLayoutResourceKindCount];
```
This will be also a lot of memory waste if it will be mostly unused.

ArielG-NV replied:
I agree. This can be reduced to `UsedRanges usedResourceRangesInputAttachmentIndex` in all likelihood since we know this resourceKind is only used for images (which have a specific layout resource kind used).

as a note: we do need this `UsedRanges` structure at a minimum though since this is just a bare minimum data structure to check ranges mapped by objects, and we are trying to ensure conflicts between Input_Attachment_Index's are caught.

I will do this in the next commit.

jkwak-work reviewed (commented):
I haven't seen the testing code yet.

jkwak-work commented on tests/glsl-intrinsic/fragment-processing/fragment-processing-interpolate-simple.slang:
```diff
@@ -0,0 +1,27 @@
+{
+    float tmpStore = inDataV1;
+    return true
+        && interpolateAtCentroid(inDataV1) != -1.0f
```
If I understood correctly, "inDataV1" is a uniform variable which means its value will be same for every pixels.
The values cannot be interpolated in this case and this will not be a proper testing for the centroid feature.

ArielG-NV replied:
inDataV1 is a fragment `input` variable. This is distinctly not a uniform, it is the interpolated output from the vertex-shader->fragment-shader.

Interpolation functions require to be used with fragment-shader `input`'s.

the value of the `input` variables rely on a vertex shader output (and modifier of the `input`, like `flat`), they may be the same for everything, they may not. We don't know the value without a vertex shader.

jkwak-work replied:
Oh. You are right. My bad. This is not a uniform.

jkwak-work commented on tests/glsl-intrinsic/fragment-processing/fragment-processing.slang:
```diff
@@ -0,0 +1,143 @@
+// CHECK_SPV-DAG: OpFwidthFine
+// CHECK_GLSL-DAG: fwidthFine
+// CHECK_SPV-DAG: OpFwidthCoarse
+// CHECK_GLSL-DAG: fwidthCoarse
```
Isn't the order deterministic?
Can you remove "-DAG"?

ArielG-NV replied:
you are correct, this should not be `-DAG` (some previous `interpolation` function fixes caused non-deterministic code order, I forgotten to change this back, thank you for catching this)

jkwak-work commented on tests/glsl-intrinsic/shader-invocation-control/shader-invocation-control.slang:
```diff
@@ -0,0 +1,40 @@
+bool testBarrier()
+{
+    raceCondition.data[0] = 3;
+    barrier();
```
I thought "barrier()" is for "groupshared" variables not the buffers.
Is it not the case?

ArielG-NV replied:
you are correct, my current implementation for barrier is wrong
(should be OpControlBarrier Workgroup Workgroup AcquireRelease|WorkgroupMemory

jkwak-work commented on tests/glsl-intrinsic/shader-invocation-group/shader-invocation-group.slang:
```diff
@@ -0,0 +1,54 @@
+{
+    return true
+        && allInvocations(true)
+        && allInvocations(false) == false
```
You can also do
```
allInvocations(gl_GlobalInvocationID.x == 0) == false
```

ArielG-NV replied:
I will replace the `==false` test for uniformity of tests as suggested

jkwak-work commented on tests/glsl-intrinsic/shader-invocation-group/shader-invocation-group.slang:
```diff
@@ -0,0 +1,54 @@
+void computeMain()
+{
+    outputBuffer.data[0] = true
+        && testInvocations()
```
Can we inline "testInvocations()" here like:
```
void computeMain()
{
    outputBuffer.data[0] = true
        && testAnyInvocation()
        && testAllInvocations()
        && testAllInvocationsEqual()
        ;
```

ArielG-NV replied:
that work, will do 👍

jkwak-work commented on tests/glsl-intrinsic/shader-memory-control/shader-memory-control.slang:
```diff
@@ -0,0 +1,100 @@
+    memoryBarrier();
+    return true
+        && raceCondition.data[0] == 1
+        ;
```
This doesn't seem effective enough.
How about something like this:
```
raceCondition.data[gl_GlobalInvocationID.x] = gl_GlobalInvocationID.x;
memoryBarrier();
if (gl_GlobalInvocationID.x == 0)
{
  for (int i = 0; i < 4; ++i)
  {
    if (raceCondition[i] != i) return false;
```

jkwak-work commented on tests/glsl-intrinsic/subpass-input/subpass-input.slang:
```diff
@@ -0,0 +1,69 @@
+//TEST:SIMPLE(filecheck=CHECK_GLSL): -target glsl -stage fragment -entry main -allow-glsl
```
I don't think we want to test GLSL for subpass feature, do we?
The document says,
```
Subpass-input functions are only available when targeting a Vulkan fragment stage.
```

ArielG-NV replied:
We do want to target subpass-input since Slang supports only Vulkan, so there is no downside to having a GLSL & SPIR-V test

jkwak-work reviewed (commented):
I left more comments including the testing code.

ArielG-NV committed:
adjustments based on review

added glsl shared modifier since it was missing

change InterpolatAtOffset to not use the glsl450 magic number. Change `barrier` to follow how glSlang works in most ways. We do not copy the atomicStorageMemory semantics used since these are invalid for vulkan through requiring a invalid OpCapability of `AtomicStorage` (which glslang does not enable, which is how it compiles)

add to tests looping to better test race conditions as suggested in review

fix syntax as per review comments

adjust comments

remove uneeded white-spaces

ArielG-NV committed:
reduce usage of memory by assuming input_Attachment_index is only used for subpass images

ArielG-NV committed:
add a constant and remove boolean flag for AttachmentIndex as per review suggestion

ArielG-NV committed:
remove uneeded white space

ArielG-NV committed:
add not uniform attribute to non uniform return invocation funcitons

jkwak-work commented on slang.h:
```diff
@@ -2144,11 +2144,12 @@ extern "C"
+        SLANG_STRUCTURED_BUFFER             = 0x07,
+        SLANG_BYTE_ADDRESS_BUFFER           = 0x08,
+        SLANG_RESOURCE_UNKNOWN              = 0x09,
+        SLANG_ACCELERATION_STRUCTURE        = 0x0A,
```
I am worried that this may break the backward compatibility.

jkwak-work commented on source/slang/slang-ast-modifier.h:
```diff
@@ -243,6 +243,11 @@ class GLSLLocationLayoutModifier : public GLSLParsedLayoutModifier
     SLANG_AST_CLASS(GLSLLocationLayoutModifier)
 };
 
+class GLSLInputAttachmentIndexModifier : public GLSLParsedLayoutModifier
```
The name should have "Layout".
Perhaps you can change it to
```
GLSLInputAttachmentLayoutModifier or
GLSLInputAttachmentIndexLayoutModifier
```

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -329,11 +329,18 @@ struct UsedRanges
     }
 };
 
+struct ParameterAndLayoutConstants
```
This is probably better with "namespace".

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -329,11 +329,18 @@ struct UsedRanges
 
+struct ParameterAndLayoutConstants
+{
+    static const uint32_t kInputAttachmentIndexUnused = 0xFFFFFFFF;
```
probably you should make this
```
static const constexpr
```
Don't instantiate a static member variable inside of struct declaration.
You need to either make it constexpr or instantiate it outside of struct declaration.

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -351,6 +358,8 @@ struct UsedRangeSet : RefObject
     // Information on what ranges of "registers" have already
     // been claimed, for each resource type
     UsedRanges usedResourceRanges[kLayoutResourceKindCount];
+    //Subpass is the only current use of InputAttachmentIndex
+    UsedRanges usedResourceRangesInputAttachmentIndex;
```
If you can increase kLayoutResourceKindCount by 1, you wouldn't need to explicitly have this new member variable.
It looks like you may be able to add one enum to "SlangParameterCategory".

ArielG-NV replied:
usedResourceRanges ensures we do not conflict on a binding, we still need to ensure we do not conflict on a InputAttachmentIndex, which is a completely different type of layout input qualifer. 

If we add another SlangParameterCategory, Subpass will need to use that category with subpassInput, else it will be unused despite us still having a subpassInput which is not currently how the code manages these categories. This would also make the reflectionAPI confusing since we have a object that satisfies a descriptor category & subpass category, but we only use the descriptor category, and have an unused subpass category...

jkwak-work commented on source/slang/slang-parameter-binding.cpp:
```diff
@@ -329,11 +329,18 @@ struct UsedRanges
+    uint32_t            inputAttachmentIndex = ParameterAndLayoutConstants::kInputAttachmentIndexUnused;
 };
 
 struct ParameterBindingAndKindInfo : ParameterBindingInfo
```
I am not convinced that you need to add a new member variable here.
Because when I look at ParameterBindingAndKindInfo, your new member variable will be added for all of the kinds.
Is it what you need?
Shouldn't you just add a new kind to "SlangParameterCategory" and use the existing member variables?

ArielG-NV replied:
We need a new member or some form of extension value to handle subpass's since we have the unique case of 3 parameter binding information's which may conflict.

Subpass's are composed of a `set`+`binding`+`input_attachment_index`. Every single other global parameter type has either: `location`+`index`, `location`+`component`, `binding`+`set`, or some variation, allowing `space`+`index` to handle all cases. This is not the case for Subpass.

jkwak-work reviewed (commented):
I left a few comments.

ArielG-NV committed:
Merge branch 'master' of https://github.com/ArielG-NV/slang into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
adjust tests to work properly with merge

ArielG-NV committed:
add image tests

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
adjust as per review comments

mostly code structure and ensuring backwards compatability

jkwak-work commented on prelude/slang-cpp-types.h:
```diff
@@ -220,10 +220,11 @@ enum
+    SLANG_STRUCTURED_BUFFER             = 0x06,
+    SLANG_BYTE_ADDRESS_BUFFER           = 0x07,
+    SLANG_RESOURCE_UNKNOWN              = 0x08,
+    SLANG_ACCELERATION_STRUCTURE        = 0x09,
```
Let's not modify these lines at all.

jkwak-work commented on slang-gfx.h:
```diff
@@ -409,6 +413,10 @@ enum class Format
     BC7_UNORM_SRGB,
 
+    R64_UINT,
+    
```
I think you can remove the empty line in the middle.

ArielG-NV replied:
This is from a different PR

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2821,6 +2821,466 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
+            {
+                OpExtension "SPV_EXT_shader_atomic_float_min_max";
+                OpCapability AtomicFloat32MinMaxEXT
+            };
```
I think this will be effectively fall-through of the switch-case.
It is fine here but I will watch out for any mistakes related to this pattern of code.

ArielG-NV replied:
I will make sure from-now on to clearly define when a case ends, I agree that I should have in the first place.

Lots of slang uses this pattern though, so I tested right now to ensure this is not a source of bugs. It does not fall through if there is any content inside a case:
```
[ForceInline] void thing()
{
    __target_switch
    {
    case glsl:
        __requireGLSLExtension("good");
        uint a = 0;
    case spirv:
        __requireGLSLExtension("bad");
    }
}
```
for GLSL, only "good" is added. 
if "bad" is in the glsl case, only "bad" is added.

```
[ForceInline] void thing()
{
    __target_switch
    {
    case spirv:
        spirv_asm{ OpCapability Tessellation };
        uint a = 0;
    case glsl:
        spirv_asm{ OpCapability Geometry };
    }
}
```
for SPIR-V only Tessellation is added, not Geometry. 
if Geometry is inside the spirv case, Geometry is added, not Tessellation.

jkwak-work replied:
Thanks for the experiment.
That is, unfortunately, little confusing behavior.

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -6803,3 +7263,728 @@ __glsl_version(430) [require(glsl)]
+        return WaveActiveAnyTrue(value);
+    }
+    }
+}
```
Can we just have "return WaveActiveAnyTrue(value)" without __target_switch?
```
public bool anyInvocation(bool value)
{
    return WaveActiveAnyTrue(value);
}
```

jkwak-work commented on source/slang/slang-ir-spirv-legalize.cpp:
```diff
@@ -247,6 +247,11 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
+                // passing as is, which needs to be a pointer (pass as is).
+                if (user->getDataType()->getOp() == kIROp_RefType
+                    && storageClass == SpvStorageClassInput)
+                    return;
```
Shouldn't this be continue or break rather than return?

ArielG-NV replied:
it should be `continue`

I will add to the test cases as well

jkwak-work commented on source/slang/slang-ir-spirv-legalize.cpp:
```diff
@@ -363,9 +368,12 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
-            // If the texture has no format decoration, try to infer it from the type.
+            // Do not infer format when shape is a subpass because SPRI-V spec defines
+            // subpass to have an unknown format 
+            if (textureType->GetBaseShape() == SLANG_TEXTURE_SUBPASS) return;
```
If I understood the comment correctly, you could separate the subpass logic from the existing logic like this:
```
        if (textureType->GetBaseShape() == SLANG_TEXTURE_SUBPASS)
        {
            // Do not infer format when shape is a subpass because SPRI-V spec defines
            // subpass to have an unknown format 
            SLANG_ASSERT(format == ImageFormat::unknown);
            return;
        }

        if (format == ImageFormat::unknown)
        {
            // If the texture has no format decoration, try to infer it from the type.
```

ArielG-NV replied:
I agree. I should also add a diagnostic error to warn users that the format should be 'unknown'

I will add this to tests.

jkwak-work commented on tests/glsl-intrinsic/fragment-processing/fragment-processing.slang:
```diff
@@ -0,0 +1,143 @@
+    vec4 fdata[2];
+    ivec4 udata[2];
+    uvec4 idata[2];
+} keepAliveBuffer;
```
Is this used? If not, can we remove it?

jkwak-work commented on tests/glsl-intrinsic/subpass-input/subpass-input.slang:
```diff
@@ -0,0 +1,69 @@
+// CHECK_SPV: OpImageRead
+// CHECK_GLSL-CHECK: subpassLoad
+    outColor = vec4(true
+        && subpassLoad(subpass) == vec4(1)
+        && subpassLoad(subpassMS, 0) == vec4(1)
+        && subpassLoad(isubpass) == ivec4(1)
+        && subpassLoad(isubpassMS, 0) == ivec4(1)
+        && subpassLoad(usubpass) == uvec4(1)
+        && subpassLoad(usubpassMS, 0) == uvec4(1)
```
Where does the value 1 come from?
My understanding is that subpassLoad is like a imageLoad.
I don't see any initialization on the XsubpassXX variables.

ArielG-NV replied:
subpass value comes from a renderpass's renderTarget, It's value will come from our testing initialization code. Currently tests do not test fragment shaders, or support subpass inputs. 

I did not add support for testing fragment shaders either (out of scope of this issue).

[(example of a subpass being passed to a shader)](https://github.com/LunarG/VulkanSamples/blob/master/API-Samples/input_attachment/input_attachment.cpp)

jkwak-work reviewed (commented):
I have a few minor comments and questions.

swoods-nv committed:
Update README.md with tables for platform & target support (#3847)

Adds tables to README.md for quick reference of which platforms and targets are supported. Links to detailed information in user guide.

Bug Id: shader-slang/slang #3667

ArielG-NV committed:
remove incorrect unneeded line changes.

adjusted syntax and removed redundant variables.

added error checking for invalid format use with SubpassInput.

set default for subpass properly to fix regression with images

ArielG-NV committed:
minor styling

use size_t and UInt for ParameterAndLayout info/constants for consistency.

csyonghe committed:
Fix lookup to prevent finding `typedef` itself. (#3848)

csyonghe committed:
Allow var/param names to be the same as type name. (#3850)

jkwak-work commented on source/slang/slang-ir-spirv-legalize.cpp:
```diff
@@ -363,9 +368,17 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         {
-            // If the texture has no format decoration, try to infer it from the type.
+            // Do not infer format when shape is a subpass because SPRI-V spec defines
+            // subpass to have an unknown format 
```
This comment may need to move.

ArielG-NV replied:
I will remove the comment since we have an error message slang throws in-place instead

jkwak-work commented on tests/glsl-intrinsic/subpass-input/input-attachment-index-use-error.slang:
```diff
@@ -0,0 +1,18 @@
+    vec4 fdata[2];
+    ivec4 udata[2];
+    uvec4 idata[2];
+} keepAliveBuffer;
```
Is this needed? "keepAliveBuffer"

jkwak-work reviewed (commented):
Only two more comments.

ArielG-NV committed:
remove a mistake `MyBlockName` accidently added

remove redundant comment

kaizhangNV committed:
Fix incorrect SPV stride for unsized array (#3837)

* Fix incorrect SPV stride for unsized array (#3825)

In '-emit-spirv-directly' mode, slang generates the stride 0
for unsized array in `OpDecorate` instructions.

For unsized array, the stride is invalid, but we need to provide
a non-zero value to pass the spirv validator.

* Decorate struct with unsized array field as 'Block'

For the struct having unsized array fields, it has to be decorated
as "Block", otherwise it will fails the spirv-val.

So we add a check at in 'emitGlobalInst' when emitting spirv for
'kIROp_StructType', where if there is unsized array field inside
the struct, emit a decorate instruction for above purpose.

* Update decoration for kIROp_SizeAndAlignmentDecoration

When add a decoration node for kIROp_SizeAndAlignmentDecoration,
we implicitly convert the 64 bit size to 32 bit. In most cases, this
should not be a problem because we won't have that large data type.
However, we use 64-bit -1 to represent the size of unsized-array,
so in that case, the conversion will change the size to 0, which is
incorrect. So change that decoration to use 64-bit size.

---------

Co-authored-by: Yong He <yonghe@outlook.com>

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
remove  "shared" modifier usage for shared memory since these will no longer work after the following change, since the name conflict is an issue.

ArielG-NV committed:
Merge branch 'glsl.meta.slang_miscGLSL_#3692' of https://github.com/ArielG-NV/slang into glsl.meta.slang_miscGLSL_#3692

ArielG-NV committed:
use groupshared in another glsl.meta test instead of "shared"

ArielG-NV committed:
set unused value to 2^32
